crates/bevy_core_pipeline/src/dof/mod.rs

```diff
@@ -0,0 +1,872 @@
+    focal_length: f32,
+
+    /// The premultiplied factor that we scale the circle of confusion by.
+    ///
+    /// This is calculated as `focal_length² / (sensor_height *
+    /// aperture_f_stops)`.
+    coc_scale_factor: f32,
+
+    /// The maximum circle of confusion diameter in pixels. See the comment in
+    /// [`DepthOfFieldSettings`] for more information.
+    max_circle_of_confusion_diameter: f32,
+
+    /// The depth value that we clamp distant objects to. See the comment in
+    /// [`DepthOfFieldSettings`] for more information.
+    max_depth: f32,
+}
+
+/// A key that uniquely identifies depth of field pipelines.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct DepthOfFieldPipelineKey {
+    /// Whether we're doing Gaussian or bokeh blur.
+    pass: DofPass,
+    /// Whether we're using HDR.
+    hdr: bool,
+    /// Whether the render target is multisampled.
+    multisample: bool,
+}
+
+/// Identifies a specific depth of field render pass.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+enum DofPass {
+    /// The first, horizontal, Gaussian blur pass.
+    GaussianHorizontal,
+    /// The second, vertical, Gaussian blur pass.
+    GaussianVertical,
+    /// The first bokeh pass: vertical and diagonal.
+    BokehPass0,
+    /// The second bokeh pass: two diagonals.
+    BokehPass1,
+}
+
+impl Plugin for DepthOfFieldPlugin {
+    fn build(&self, app: &mut App) {
+        load_internal_asset!(app, DOF_SHADER_HANDLE, "dof.wgsl", Shader::from_wgsl);
+
+        app.add_plugins(ExtractComponentPlugin::<DepthOfFieldSettings>::default())
+            .add_plugins(UniformComponentPlugin::<DepthOfFieldUniform>::default());
+
+        let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
+            return;
+        };
+
+        render_app
+            .init_resource::<SpecializedRenderPipelines<DepthOfFieldPipeline>>()
+            .init_resource::<DepthOfFieldGlobalBindGroup>()
+            .add_systems(
+                Render,
+                (
+                    configure_depth_of_field_view_targets,
+                    prepare_auxiliary_depth_of_field_textures,
+                )
+                    .after(prepare_view_targets)
+                    .in_set(RenderSet::ManageViews),
+            )
+            .add_systems(
+                Render,
+                (
+                    prepare_depth_of_field_view_bind_group_layouts,
+                    prepare_depth_of_field_pipelines,
+                )
+                    .chain()
+                    .in_set(RenderSet::Prepare),
+            )
+            .add_systems(
+                Render,
+                prepare_depth_of_field_global_bind_group.in_set(RenderSet::PrepareBindGroups),
+            )
+            .add_render_graph_node::<ViewNodeRunner<DepthOfFieldNode>>(Core3d, Node3d::DepthOfField)
+            .add_render_graph_edges(
+                Core3d,
+                (Node3d::Bloom, Node3d::DepthOfField, Node3d::Tonemapping),
```

Comment:
Reminder for myself to check that this ordering is sensible with respect to TAA and SSAO.

Reply:
I think this order will blur the bloom fringes of focused objects if the underlying depth values are out of focus, will have to test.